cmd/gateway-azure.go

```diff
@@ -108,21 +100,15 @@ func s3MetaToAzureProperties(s3Metadata map[string]string) (storage.BlobMetadata
 // `.GetMetadata()` lower-cases all header keys, so this is taken into
 // account by this function.
 func azurePropertiesToS3Meta(meta storage.BlobMetadata, props storage.BlobProperties) map[string]string {
-	// Remap underscores to hyphens to restore encryption
-	// headers. See s3MetaToAzureProperties for details.
-	gatewayHeaders := map[string]string{
-		"X-Amz-Meta-x_minio_key":     "X-Amz-Meta-X-Amz-Key",
-		"X-Amz-Meta-x_minio_matdesc": "X-Amz-Meta-X-Amz-Matdesc",
-		"X-Amz-Meta-x_minio_iv":      "X-Amz-Meta-X-Amz-Iv",
-	}
-
 	s3Metadata := make(map[string]string)
 	for k, v := range meta {
 		// k's `x-ms-meta-` prefix is already stripped by
 		// Azure SDK, so we add the AMZ prefix.
 		k = "X-Amz-Meta-" + k
-		if nk, ok := gatewayHeaders[k]; ok {
-			k = nk
+		if strings.Contains(k, "__") {
+			k = strings.Replace(k, "__", "_", -1)
+		} else {
+			k = strings.Replace(k, "_", "-", -1)
```

Comment:
This won't work for case like `My_Metadata__Name` which is expected to be `My-Metadata_name`